### PR TITLE
[Segment Replication] Add Segment Replication backpressure rejection stats to _nodes/stats

### LINKED ISSUE #10501: [Remote Store] Include SegRep and Remote Store Rejections in  node/stats api
**Describe the bug**
 We reject write requests when replicas are lagging or remote store segment store is lagging - https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java#L534

We should include this rejections in our node stats in a separate blob of its own on the lines of shard_indexing_pressure/search backpressure 

Proposed blobs are 

```
      "remote_store_pressure" : {
       "rejections" : count,
      },
      "segment_replication_pressure" : {
       "rejections" : count,
      },

```

**Alternate**

We can have one blob as well to account for both of them

```
      "remote_store_pressure" : {
       "upload_rejections" : count,
"lag_rejections" : count
      },
```

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
Add Segment Replication  backpressure rejection stats to _nodes/stats

### Related Issues
Resolves #10501 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodeStats.java [hunk 1]
```diff
@@ -128,6 +129,9 @@ public class NodeStats extends BaseNodeResponse implements ToXContentFragment {
     @Nullable
     private SearchBackpressureStats searchBackpressureStats;
 
+    @Nullable
+    private SegmentReplicationStats segmentReplicationStats;
```
Comment (mch2) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodeStats.java:133]: We are only displaying the back pressure rejection count here and don't need to compute everything this object has. Let's make a new segmentreplicationRejectionStats object instead? We can return that right from the tracker.

**Review by mch2 — COMMENTED** [review]
This needs a test pls

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodeStats.java (hunk 1 above)
Comment (Rishikesh1159) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodeStats.java]: Added in new commit

**Comment by Rishikesh1159** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28165/
> * **CommitID:** [317ec74](https://github.com/opensearch-project/OpenSearch/commit/317ec74cc8e0b34fab03048f33690956627dcde5)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

flaky test: https://github.com/opensearch-project/OpenSearch/issues/10643

File: server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java [hunk 2]
```diff
@@ -28,11 +29,24 @@ public class SegmentReplicationStatsTracker {
     private final IndicesService indicesService;
     private final Map<ShardId, AtomicInteger> rejectionCount;
 
+    @Inject
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java:32]: An instance of this class is created in Node and passed to other components.  We should bind to that instance only and remove this.

File: server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java [hunk 3]
```diff
@@ -28,11 +29,24 @@ public class SegmentReplicationStatsTracker {
...
     public SegmentReplicationStatsTracker(IndicesService indicesService) {
         this.indicesService = indicesService;
         rejectionCount = ConcurrentCollections.newConcurrentMap();
     }
 
+    public SegmentReplicationRejectionStats getRejectionStats() {
+        long rejectionCount = 0;
+        for (IndexService indexService : indicesService) {
+            for (IndexShard indexShard : indexService) {
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java:41]: lets read the isSegRepEnabled setting off indexService before iterating all shards.

File: server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java (hunk 2 above)
Comment (Rishikesh1159) [reply server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java]: Added in latest commit

File: server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java (hunk 3 above)
Comment (Rishikesh1159) [reply server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java]: Added in latest commit

File: server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java [hunk 4]
```diff
@@ -33,6 +33,20 @@ public SegmentReplicationStatsTracker(IndicesService indicesService) {
         rejectionCount = ConcurrentCollections.newConcurrentMap();
     }
 
+    public SegmentReplicationRejectionStats getRejectionStats() {
+        long rejectionCount = 0;
+        for (IndexService indexService : indicesService) {
+            if (indexService.getIndexSettings().isSegRepEnabled()) {
+                for (IndexShard indexShard : indexService) {
+                    if (indexShard.indexSettings().isSegRepEnabled() && indexShard.routingEntry().primary()) {
+                        rejectionCount += getStatsForShard(indexShard).getRejectedRequestCount();
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/SegmentReplicationStatsTracker.java:42]: `getStatsForShard` still fetches all per group replication stats that we don't need here.  We can get the rejection counts directly from this class's `rejectionCount` map.

we also don't need a per shard isSegRepEnabled check, the values in the rejectionCount map will only get populated for SR indices.  You can directly sum the values of the map.

```java
    public SegmentReplicationRejectionStats getTotalRejectionCount() {
        return new SegmentReplicationRejectionStats (this.rejectionCount.values().stream().mapToInt(AtomicInteger::get).sum());
    }
```

File: server/src/main/java/org/opensearch/index/SegmentReplicationRejectionStats.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentFragment;
+import org.opensearch.core.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+/**
+ * Segment Replication Rejection Stats.
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/index/SegmentReplicationRejectionStats.java:21]:
```suggestion
 * Segment replication rejection stats.
```

File: server/src/test/java/org/opensearch/index/SegmentReplicationStatsTrackerTests.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import org.opensearch.indices.IndicesService;
+import org.opensearch.test.OpenSearchTestCase;
+
+import static org.mockito.Mockito.mock;
+
+public class SegmentReplicationStatsTrackerTests extends OpenSearchTestCase {
+
+    private IndicesService indicesService = mock(IndicesService.class);
+
+    public void testRejectedCountWhenEmpty() {
```
Comment (dreamer-89) [inline server/src/test/java/org/opensearch/index/SegmentReplicationStatsTrackerTests.java:20]: Do we need coverage for non-zero rejection count use case ?

File: server/src/main/java/org/opensearch/index/SegmentReplicationRejectionStats.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @opensearch.internal
+ */
+public class SegmentReplicationRejectionStats implements Writeable, ToXContentFragment {
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/index/SegmentReplicationRejectionStats.java:25]: May be we can remove this class as we are only tracking a long variable here unless we seek future changes to add more fields.

Comment (Rishikesh1159) [reply server/src/main/java/org/opensearch/index/SegmentReplicationRejectionStats.java:25]: Yes, @mch2 suggested [here](https://github.com/opensearch-project/OpenSearch/pull/10656#discussion_r1361614250) for a new class

Comment (mch2) [reply server/src/main/java/org/opensearch/index/SegmentReplicationRejectionStats.java:25]: Would rather we keep this so we aren't directly writing the xcontent in NodeStats

File: server/src/test/java/org/opensearch/index/SegmentReplicationStatsTrackerTests.java (hunk 6 above)
Comment (Rishikesh1159) [reply server/src/test/java/org/opensearch/index/SegmentReplicationStatsTrackerTests.java]: Added in new commit

**Review by mch2 — APPROVED** [review]

**Comment by Rishikesh1159** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28605/
> * **CommitID:** [3cf7f33](https://github.com/opensearch-project/OpenSearch/commit/3cf7f33e3610576e346b8f59fb0addafb11c2065)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

known flaky test: https://github.com/opensearch-project/OpenSearch/issues/10006

**Comment by Rishikesh1159** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28611/
> * **CommitID:** [3cf7f33](https://github.com/opensearch-project/OpenSearch/commit/3cf7f33e3610576e346b8f59fb0addafb11c2065)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

known flaky test: https://github.com/opensearch-project/OpenSearch/issues/10709

**Pull request merged**